Deny global_asm in deny(unsafe_code)

gimbles opened the pull request:
Fixes #103078

lcnr committed:
make `Sized` coinductive

lcnr committed:
add some more tests

Urgau committed:
Test that target feature mix up with homogeneous floats is sound

This is basically is ripoff of src/test/ui/simd/target-feature-mixup.rs
but for floats and without #[repr(simd)]

BGR360 committed:
Allow specialized const trait impls.

Fixes #95186.
Fixes #95187.

WaffleLapkin committed:
recover wrong-cased `use`s (`Use`, `USE`, etc)

JakobDegen committed:
Add support for custom MIR parsing

compiler-errors committed:
bless tests

estebank committed:
Tweak span for `#[must_use]`

Do not point at whole statement, only at the expression (skip pointing at `;`)

estebank committed:
Consider `#[must_use]` annotation on `async fn` as also affecting the `Future::Output`

No longer lint against `#[must_use] async fn foo()`.

When encountering a statement that awaits on a `Future`, check if the
`Future`'s parent item is annotated with `#[must_use]` and emit a lint
if so. This effectively makes `must_use` an annotation on the
`Future::Output` instead of only the `Future` itself.

Fix #78149.

estebank committed:
review comments

SLASHLogin committed:
Port unknown feature diagnostic to the new framework

SLASHLogin committed:
Formatting

SLASHLogin committed:
locales formatting

SLASHLogin committed:
Trailing whitespaces

SLASHLogin committed:
Change String in structs to &'a str

SLASHLogin committed:
Missing lifetime parameter and formatting

SLASHLogin committed:
Import `error creating import library`

SLASHLogin committed:
Port Instrument coverage requires llvm 12 to the new struct

SLASHLogin committed:
Port `symbol_already_defined` error

SLASHLogin committed:
Fix diag() and formatting

SLASHLogin committed:
Port branch protection on aarch64

SLASHLogin committed:
Formatting

SLASHLogin committed:
Port layout size overflow

SLASHLogin committed:
Port InvalidMinimumAlignment

SLASHLogin committed:
Port LinkageConstOrMutType error

SLASHLogin committed:
Reuse SymbolAlreadyDefined

SLASHLogin committed:
Port SanitizerMemtagRequiresMte

SLASHLogin committed:
Port ArchiveBuildFailure

SLASHLogin committed:
Import ErrorWritingDEFFile

SLASHLogin committed:
Port ErrorCallingDllTool

SLASHLogin committed:
Port `DlltoolFailImportLibrary` and implement `IntoDiagnosticArg` for `Cow<'a, str>`

SLASHLogin committed:
Port `UnknownArchiveKind`

SLASHLogin committed:
Port `MissingFeatures` and `TargetFeatureDisableOrEnable`

SLASHLogin committed:
Fix CI
Add missing 'the' to the error en-US translation

SLASHLogin committed:
Correct tests to match errors

WaffleLapkin committed:
Recover wrong cased keywords starting functions

BGR360 committed:
Disallow specializing on const impls with non-const impls.

AndyJado committed:
lint auto pass

Revert "lint auto pass"

This reverts commit e58e4466384924c491a932d3f18ef50ffa5a5065.

WaffleLapkin committed:
Replace some `bool` params with an enum

zhaixiaojuan committed:
Add loongarch64 abi support

clubby789 committed:
Prevent auto-application of associated functions with placeholders

clubby789 committed:
Allow inferring generic arguments for associated methods

clubby789 committed:
Update UI test

clubby789 committed:
Use `FmtPrinter` instead of creating `Instance`

clubby789 committed:
Attempt to fix arguments of associated functions

alexcrichton committed:
Implement the `+whole-archive` modifier for `wasm-ld`

This implements the `Linker::{link_whole_staticlib,link_whole_rlib}`
methods for the `WasmLd` linker used on wasm targets. Previously these
methods were noops since I think historically `wasm-ld` did not have
support for `--whole-archive` but nowadays it does, so the flags are
passed through.

clubby789 committed:
Split non-fixable case to different test

clubby789 committed:
Replace boxed iterator with vec collect

jyn514 committed:
More build-manifest docs

jyn514 committed:
More build-manifest docs

jyn514 committed:
build-manifest: Add a macro that makes it impossible to typo `-preview`, or have a mismatch between parsing and stringifying

jyn514 committed:
Use `PkgType` to determine which packages to add to the manifest

Previously, these had to be hard-coded (i.e. specified in both `PkgType` and `fn package`). Now they only have to be specified in `PkgType`.

jyn514 committed:
Use `PkgType::is_preview` to determine whether to add a rename to the manifest

This caught a missing preview rename for `llvm-tools`.

jyn514 committed:
Use an exhaustive match in `target_host_combination`.

This avoids bugs where components are added to one part of the manifest but not another.

feniljain committed:
feat: add config for inserting must_use in `generate_enum_as_method`

compiler-errors committed:
DiagnosticBuilder -> Diagnostic

compiler-errors committed:
Make span_suggestions take IntoIterator

DropDemBits committed:
Migrate most of `ide_assists::utils` to format arg capture

clubby789 committed:
Update UI test

DropDemBits committed:
Migrate assists to format args captures, part 1

Commit:
rustc_codegen_ssa: Better code generation for niche discriminants.

In some cases we can avoid arithmetic before checking whether a niche
represents an untagged variant.

This is relevant to #101872

DropDemBits committed:
Migrate assists to format args captures, part 3

jackh726 committed:
Add an optional Span to BrAnon and use it to print better error for HRTB error from generator interior

joboet committed:
std: remove lock wrappers in `sys_common`

joboet committed:
std: fix double-free of mutex

jackh726 committed:
Get spans for a couple more region types, add some optimizations, and extend test

notJoon committed:
fix broken links in guide.md

Commit:
Add context to compiler error message

Changed `creates a temporary which is freed while still in use` to
`creates a temporary value which is freed while still in use`

Commit:
Update tests to match error message changes

wesleywiser committed:
Fix Access Violation when using lld & ThinLTO on windows-msvc

Users report an AV at runtime of the compiled binary when using lld and
ThinLTO on windows-msvc. The AV occurs when accessing a static value
which is defined in one crate but used in another. Based on the
disassembly of the cross-crate use, it appears that the use is not
correctly linked with the definition and is instead assigned a garbage
pointer value.

If we look at the symbol tables for each crates' obj file, we can see
what is happening:

*lib.obj*:

```
COFF SYMBOL TABLE
...
00E 00000000 SECT2  notype       External     | _ZN10reproducer7memrchr2FN17h612b61ca0e168901E
...
```

*bin.obj*:

```
COFF SYMBOL TABLE
...
010 00000000 UNDEF  notype       External     | __imp__ZN10reproducer7memrchr2FN17h612b61ca0e168901E
...
```

The use of the symbol has the "import" style symbol name but the
declaration doesn't generate any symbol with the same name. As a result,
linking the files generates a warning from lld:

> rust-lld: warning: bin.obj: locally defined symbol imported: reproducer::memrchr::FN::h612b61ca0e168901 (defined in lib.obj) [LNK4217]

and the symbol reference remains undefined at runtime leading to the AV.

To fix this, we just need to detect that we are performing ThinLTO (and
thus, static linking) and omit the `dllimport` attribute on the extern
item in LLVM IR.

justinmmott committed:
addressed https://github.com/rust-lang/rust-analyzer/issues/12536

emilio committed:
ide: Generate monikers for local crates.

fmease committed:
test attr: point at return type if Termination bound unsatisfied

BlackHoleFox committed:
Merge apple_base and apple_sdk_base into one module

BlackHoleFox committed:
Cleanup Apple target specifications

chenyukang committed:
fix #103435, unused lint won't produce invalid code

chenyukang committed:
fix parentheses surrounding spacing issue in parser

feniljain committed:
refactor: remove repetitive string interpolation and doc changes

chenyukang committed:
Fix #103451, find_width_of_character_at_span return width with 1 when reaching end

cjgillot committed:
Tweak signatures in rustc_middle::hir::map.

cjgillot committed:
Resolve lifetimes independently for each item-like.

cjgillot committed:
Resolve lifetimes using the regular logic for RPIT.

chenyukang committed:
suggest calling the method of the same name when method not found

wesleywiser committed:
Add test case

chenyukang committed:
add testcase for suggest self

AndyJado committed:
first move on a nested span_label

Apply suggestions from code review

Co-authored-by: David Wood <agile.lion3441@fuligin.ink>

Commit:
stabilize `int_log`

chenyukang committed:
fix #103587, Recover from common if let syntax mistakes/typos

jyn514 committed:
Use PkgType in more places

In particular, this avoids serializing and parsing the pkg to a string,
which allows getting rid of `PkgType::Other` altogether

jyn514 committed:
Use PkgType for profiles, too

This makes it easier to remove `is_preview` from components in the future if we choose to do so.

Noratrieb committed:
Store `ErrorGuaranteed` in `ErrorReported`

Noratrieb committed:
Small parser cleanups

Noratrieb committed:
Add `Tracker` to track matching operations

This should allow us to collect detailed information without slowing
down the inital hot path.

Noratrieb committed:
Factor out matching into `try_match_macro`

This moves out the matching part of expansion into a new function. This
function will try to match the macro and return an error if it failed to
match. A tracker can be used to get more information about the matching.

Noratrieb committed:
Retry matching with tracking for diagnostics

For now, we only collect the small info for the `best_failure`, but
using this tracker, we can easily extend it in the future to track
things with more performance overhead.

We cannot retry cases where the macro failed with a parser error that
was emitted already, as that would cause us to emit the same error to
the user twice.

Noratrieb committed:
Add some debug logs to macro matching

These were useful while debugging, so I'll leave them here.

Commit:
remove redundent "<>" for ty::Slice with reference type
this fix #103271

WaffleLapkin committed:
Add examples for `pointer::mask`

koka831 committed:
fix: async trait method for `unnecessary_async`

CastilloDel committed:
Reduce the scope of allow(rustc::potential_query_instability) in rustc_trait_selection

Make InferCtxtExt use a FxIndexMap

This should be faster, because the map is only being used to iterate,
which is supposed to be faster with the IndexMap

Make the user_computed_preds use an IndexMap

It is being used mostly for iteration, so the change shouldn't result in
a perf hit

Make the RegionDeps fields use an IndexMap

This change could be a perf hit. Both `larger` and `smaller` are used
for iteration, but they are also used for insertions.

Make types_without_default_bounds use an IndexMap

It uses extend, but it also iterates and removes items. Not sure if
this will be a perf hit.

Make InferTtxt.reported_trait_errors use an IndexMap

This change brought a lot of other changes. The map seems to have been
mostly used for iteration, so the performance shouldn't suffer.

Add FIXME to change ProvisionalEvaluationCache.map to use an IndexMap

Right now this results in a perf hit. IndexMap doesn't have
the `drain_filter` API, so in `on_completion` we now need to iterate two
times over the map.

RalfJung committed:
teach ./miri how to do Josh syncs

RalfJung committed:
merge rustup-toolchain into ./miri

RalfJung committed:
explain how to go back to rustup-managed Miri

RalfJung committed:
Stacked Borrows: make scalar field retagging the default

koka831 committed:
Simplify the procedure

fix: remove unused import

unexge committed:
Record diverging match arms in `InferenceResult`

unexge committed:
Update auto generated tests

beepster4096 committed:
impl condvars for windows

RalfJung committed:
rustup

feniljain committed:
fix: make custom expr prefix completions to understand refs

SLASHLogin committed:
Flatten diagnostic structs

SLASHLogin committed:
Port diagnostics created by `Handler`

SLASHLogin committed:
Simplify existing Diagnostic implementations

jyn514 committed:
Remove unnecessary argument to `package`

beepster4096 committed:
use enum for condvar locks

beepster4096 committed:
fix shared behavior and add tests

beepster4096 committed:
add acquire when init once is already complete

Rageking8 committed:
fix dupe word typos

Rageking8 committed:
followup for pr 2640

lowr committed:
fix: disregard type variable expectation for if expressions

compiler-errors committed:
Check for substs compatibility for RPITITs

compiler-errors committed:
Remap RPITIT substs properly

compiler-errors committed:
Fix ICE in default impl error reporting

Commit:
Create `Callable`s for generic types implementing `FnOnce`

Commit:
Revert "Record diverging match arms in `InferenceResult`"

This reverts commit 319611b7382fc4c84170519dade68f4f558a44b1.

unexge committed:
Use let-else statements in `Convert to guarded return` assist

compiler-errors committed:
Use TraitEngine in more places, make FulfillmentCtxt constructor more private

RalfJung committed:
fix ./miri bench

feniljain committed:
fix: indentation after inserting `#must_use`

Co-authored-by: Lukas Wirth <lukastw97@gmail.com>

lnicola committed:
Bump ovsx

SLASHLogin committed:
Delay diagnostic translation in `rustc_codegen_ssa`

nnethercote committed:
Rename some variables.

These have been bugging me for a while.

- `literal_text`: `src` is also used and is shorter and better.
- `first_char`: used even when "first" doesn't make sense; `c` is
  shorter and better.
- `curr`: `c` is shorter and better.
- `unescaped_char`: `result` is also used and is shorter and better.
- `second_char`: these have a single use and can be elided.

nnethercote committed:
Clarify range calculations.

There is some subtlety here.

nnethercote committed:
Use `Mode` less.

It's passed to numerous places where we just need an `is_byte` bool.
Passing the bool avoids the need for some assertions.

Also rename `is_bytes()` as `is_byte()`, to better match `Mode::Byte`,
`Mode::ByteStr`, and `Mode::RawByteStr`.

nnethercote committed:
Make non-ASCII errors more consistent.

There are three kinds of "byte" literals: byte literals, byte string
literals, and raw byte string literals. None are allowed to have
non-ASCII chars in them.

Two `EscapeError` variants exist for when that constraint is violated.
- `NonAsciiCharInByte`: used for byte literals and byte string literals.
- `NonAsciiCharInByteString`: used for raw byte string literals.

As a result, the messages for raw byte string literals use different
wording, without good reason. Also, byte string literals are incorrectly
described as "byte constants" in some error messages.

This commit eliminates `NonAsciiCharInByteString` so the three cases are
handled similarly, and described correctly. The `mode` is enough to
distinguish them.

Note: Some existing error messages mention "byte constants" and some
mention "byte literals". I went with the latter here, because it's a
more correct name, as used by the Reference.

pd4d10 committed:
docs: add crates section to the manual

nnethercote committed:
Improve comments.

Remove a low-value comment, remove a duplicate comment, and correct a
third comment.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3384628976/jobs/5621807249) [(plain)](https://github.com/rust-lang/rust/commit/3235c05f39bc750a8e9c4f8782cb7ca298cae78b/checks/9269149152/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
```

</details>

Nilstrieb commented:
You have your test binary checked in

SLASHLogin committed:
Add `replace_args` method for `rustc_errors::diagnostic::Diagnostic`

SLASHLogin committed:
Formatting

PeteDevoy committed:
Fix broken link in error code E0706 docs

Corresponding subsection in async book is not `07.05` not `07.06`.

The information on the linked page is the same so it may be reasonable to remove the whole sentence.

fee1-dead reviewed (changes requested):
Please add a ui test, also please squash the commit so that the binary does not appear in the history.

lnicola committed:
Allow ovsx publishing to fail

fee1-dead commented on compiler/rustc_middle/src/lint.rs:
```diff
@@ -483,7 +483,10 @@ pub fn in_external_macro(sess: &Session, span: Span) -> bool {
         ExpnKind::AstPass(_) | ExpnKind::Desugaring(_) => true, // well, it's "external"
-        ExpnKind::Macro(MacroKind::Bang, _) => {
+        ExpnKind::Macro(MacroKind::Bang, id) => {
+            if id.as_str() == "core::arch::global_asm" {
+                return false;
+            }
```
Please add a comment explaining that this is needed such that the `unsafe_code` lint works correctly.

Also, what if an external macro generates `global_asm`? Would that be denied by the compiler?

nicholasbishop committed:
Promote {aarch64,i686,x86_64}-unknown-uefi to Tier 2

MCP: https://github.com/rust-lang/compiler-team/issues/555

Veykril committed:
Clarify what commands are debug commands in VSCode

nnethercote committed:
Refactor `cook_lexer_literal`.

It deals with eight cases: ints, floats, and the six quoted types
(char/byte/strings). For ints and floats we have an early return, and
the other six types fall through to the code at the end, which makes the
function hard to read.

This commit rearranges things to avoid the early returns.

nnethercote committed:
Inline and remove `validate_int_literal`.

It has a single callsite, and is fairly small. The `Float` match arm
already has base-specific checking inline, so this makes things more
consistent.

nnethercote committed:
Remove `unescape_byte_literal`.

It's easy to just use `unescape_literal` + `byte_from_char`.

beepster4096 committed:
refactor into private functions

beepster4096 committed:
clarify no_data_race_after_complete test

nnethercote committed:
Rename some `result` variables as `res`, for consistency.

ehuss committed:
Move intra-doc link checks to a separate function.

ehuss committed:
Remove reference from the intra-doc link checker.

chenyukang committed:
code cleanup with err.emit_unless

Noratrieb committed:
Small style improvements

AndyJado committed:
remove old var_span_path_only

doc comment

onestacked committed:
Made `Hash` and `Hasher` const_trait

SLASHLogin committed:
Use `LayoutError`'s implementation of `IntoDiagnostic`

Veykril committed:
Lower unsafety of fn pointer and fn item types

onestacked committed:
Make `BuildHasher` const_trait

Veykril committed:
Mark the Memory Usage command as debug command

Veykril committed:
fix: Fix reference searching only accounting substrings instead of whole identifiers

mati865 committed:
Update several crates for improved support of the new targets

This helps with `*-windows-gnullvm` targets

alex committed:
Remove linuxkernel targets

These are not used by the actual Rust-for-Linux project, so they're mostly just confusing.

pd4d10 committed:
docs: fix adoc links

nicholasbishop committed:
Limit efiapi calling convention to supported arches

Supported architectures in UEFI are described here:
https://uefi.org/specs/UEFI/2.10/02_Overview.html#calling-conventions

Changes to tests modeled on 8240e7aa101815e2009c7d03b33dd2566d843e73.

https://github.com/rust-lang/rust/issues/65815

compiler-errors committed:
rebase conflict

ted-tanner committed:
Place config.toml in current working directory if config not found

clubby789 committed:
Rebase and update test

compiler-errors committed:
Suggest is_some or let when encountering Option and bool type mismatch

BGR360 committed:
Add #[const_trait] where needed in tests.

BGR360 committed:
Require `~const` qualifier on trait bounds in specializing impls if present in base impl.

RalfJung committed:
bootstrap: add support for running Miri on a file

RalfJung committed:
bootstrap: put Miri sysroot into local build dir

BoxyUwU committed:
fixyfixfix

joboet committed:
update LLVM submodule to make libunwind work on SGX

onestacked committed:
Made `Sip` const `Hasher`

onestacked committed:
Added `const_hash` tracking issue id

joboet committed:
update debuginfo check

nicholasbishop committed:
Use aapcs for efiapi calling convention on arm

On arm, llvm treats the C calling convention as `aapcs` on soft-float
targets and `aapcs-vfp` on hard-float targets [1]. UEFI specifies in the
arm calling convention that floating point extensions aren't used [2],
so always translate `efiapi` to `aapcs` on arm.

[1]: https://github.com/rust-lang/compiler-builtins/issues/116#issuecomment-261057422
[2]: https://uefi.org/specs/UEFI/2.10/02_Overview.html#detailed-calling-convention

https://github.com/rust-lang/rust/issues/65815

RalfJung committed:
disable btree size tests on Miri

RalfJung committed:
run alloc benchmarks in Miri and fix UB

kamirr committed:
Add --print=split-debuginfo

This option prints all supported values for -Csplit-debuginfo=.., i.e.
only stable ones on stable/beta and all of them on nightly/dev.

RalfJung committed:
fmt

onestacked committed:
Removed unnecessary Trait bound

yancyribbens committed:
rustdoc: Add mutable to the description

yancyribbens committed:
rustdoc: Add mutable to the description

onestacked committed:
Fix `const_fn_trait_ref_impl`, add test for it

GuillaumeGomez committed:
Fix invalid background-image file name

GuillaumeGomez committed:
Migrate crate-search element to CSS variables

GuillaumeGomez committed:
Extend crate-search div GUI tests

notriddle committed:
rustdoc: refactor `notable_traits_decl` to just act on the type directly

onestacked committed:
Reworked const fn ref tests

onestacked committed:
const Compare Tuples

ted-tanner committed:
Only set config.config to None when using default path

jackh726 committed:
Add a known that this is a known limitation

notriddle committed:
rustdoc: use javascript to layout notable traits popups

Fixes #102576

notriddle committed:
Update to latest version of flate2

rejyr committed:
fix: lint against lint functions

fix: lint against the functions `LintContext::{lookup_with_diagnostics,lookup,struct_span_lint,lint}`, `TyCtxt::struct_lint_node`, `LintLevelsBuilder::struct_lint`.

notriddle committed:
rustdoc: fix font color inheritance from body, and test

nnethercote committed:
Simplify `unescape_{char,byte}`.

The `usize` isn't needed in the error case.

chenyukang committed:
use subdiagnostic for sugesting add let

emilyalbini committed:
clarify licensing situation of mpsc and spsc queue

WaffleLapkin committed:
Remove an address comparison from the parser

Ayush1325 committed:
Add retry flag to remote-test-server

This allows retrying binding TCP Socket multiple times. This is useful
when using emulators as network might not be available in the beginning.
This was orignally implemented in https://github.com/rust-lang/rust/pull/100316

Signed-off-by: Ayush Singh <ayushsingh1325@gmail.com>

GuillaumeGomez committed:
Update browser-ui-test version to 0.13.1

GuillaumeGomez committed:
Add new option to prevent CORS failures

oli-obk committed:
Simplify some `Autoderef::new` calls

oli-obk committed:
Remove overloaded_span argument from `new`, where it is usually redundant with the main span

oli-obk committed:
Remove an unused span

onestacked committed:
Test const `Hash`, fix nits

WaffleLapkin committed:
Fix outdated comment

Co-authored-by: clubby789 <jamie@hill-daniel.co.uk>

cjgillot committed:
Make AbsoluteBytePos a u64.

Manishearth committed:
Add llvm-main to triagebot.toml

SLASHLogin committed:
Update compiler/rustc_codegen_llvm/src/back/archive.rs

Co-authored-by: David Wood <agile.lion3441@fuligin.ink>

GuillaumeGomez committed:
Migrate :target rules to use CSS variables

GuillaumeGomez committed:
Add GUI test for :target

ehuss committed:
Update books

BoxyUwU committed:
comment

BoxyUwU committed:
tests

fee1-dead committed:
Use `nominal_obligations_without_const` in wf for FnDef

mucinoab committed:
Parser: Recover from using colon as path separator in imports

chenyukang committed:
Fix #104086, Tighten the 'introduce new binding' suggestion

jyn514 committed:
Fix `rustdoc --version` when used with download-rustc

Previously, rustdoc would unconditionally report the version that *rustc* was compiled with.
That showed things like `nightly-2022-10-30`, which wasn't right, since this was a `dev` build compiled from source.

Fix it by changing `rustc_driver::version` to a macro expanded at invocation time.

chenyukang committed:
fix tests and code cleanup

onestacked committed:
Cleanup fn trait ref test

chenyukang committed:
bless clippy

Noratrieb committed:
Ignore "Change InferCtxtBuilder from enter to build" in git blame

Because it changed the indentation of many things, this commit caused a lot of diff with no functional changes, so we should ignore it.

SLASHLogin committed:
Add constructor for `Diagnostic` that takes `Vec<(DiagnosticMessage, Style)>`

AndyJado committed:
var_subdiag refinement

trim old

AndyJado committed:
struct error E0505

chenyukang committed:
add 'is_assign_rhs' to avoid weird suggesting 'let'

SLASHLogin committed:
Add missing `emitted_at` to the `Diagnostic`

BoxyUwU committed:
docs

GuillaumeGomez committed:
Update to new browser-ui-test version

camsteffen committed:
Add domain size check to fix ICE

notriddle committed:
rustdoc: sort output to make it deterministic

Noratrieb committed:
Don't ICE on operator trait methods with generic methods

Emit a fatal error instead.

Noratrieb committed:
Display help message when fluent arg was referenced incorrectly

The fluent argument syntax is a little special and easy to get wrong, so
we emit a small help message when someone gets it wrong.

Example:
```
parser_mismatched_closing_delimiter = mismatched closing delimiter: `${delimiter}`
```
panics with
```
thread 'rustc' panicked at 'Encountered errors while formatting message for `parser_mismatched_closing_delimiter`
help: Argument `delimiter` exists but was not referenced correctly. Try using `{$delimiter}` instead
attr: `None`
args: `FluentArgs([("delimiter", String("}"))])`
errors: `[ResolverError(Reference(Message { id: "delimiter", attribute: None }))]`', compiler/rustc_errors/src/translation.rs:123:21
```

kubycsolutions committed:
Avoid runtime dependency on static libstdc++

Usually, we do want to use the static C++ library when building rustc_llvm, but do not want to have that dependency at compiler runtime. Change the defaults to Make It So.

compiler-errors committed:
bless a chalk test

ibraheemdev committed:
update mailmap

compiler-errors committed:
Don't ICE when encountering ConstKind::Error in RequiredConstsVisitor

compiler-errors committed:
Use const_error_with_guaranteed more

BoxyUwU committed:
broken links go brrrrr

BGR360 committed:
Apply PR feedback.

compiler-errors committed:
More nits

estebank committed:
Fix tests after rebase

BGR360 committed:
Oops, bless this test.

cjgillot committed:
Visit opaque types during type collection too.

cjgillot committed:
Clean-up formatting.

notriddle committed:
rustdoc: fix HTML validation failure by escaping `data-ty`

gimbles commented:
OH NO

gimbles commented:
I messed up a rebase I seriously didn't mean to ping everyone 😓
